### ISSUE TITLE
Table metadata includes targets for tables a user doesn't have access to

### ISSUE THREAD

**Issue by mazameli** [metabase/metabase#3867]
In v0.21.0-rc2, in Chrome, if a group's permission setting for a database is set to Limited Access, and then they are granted access to a single table in that DB, it appears that users in that group still cannot run queries against that table in the query builder, though they can select the table from the data dropdown.

Also, when trying to grant selective access to more than a single table in a DB for which a group has Limited Access, when saving the permissions changes, only the table(s) for which they have _newly_ been granted access will show up as green; if they had access to any other table(s), that access gets revoked.

Video:

[permissions-bugs.mov.zip](https://github.com/metabase/metabase/files/632503/permissions-bugs.mov.zip)

Only seeing one JS error:
<img width="1026" alt="screen shot 2016-12-05 at 2 52 47 pm" src="https://cloud.githubusercontent.com/assets/2223916/20905880/8c636d64-bafa-11e6-83d6-759c8e16605e.png">

**Comment by tlrobinson** [comment]
This appears to be due to foreign key relationships from a table the user has access to to a table they don't have access to. The QB fetches the source table, which they user has access to, but then tries to fetch any tables accessible through it's foreign keys.

I think I would expect for the backend to not return the `target` field in the metadata in this case, since the user doesn't have access to that field. i.e. they should see the fk, but not the pk it points to.

@salsakran @camsaul Thoughts?

**Comment by tlrobinson** [comment]
Regarding the second half ("only the table(s) for which they have newly been granted access will show up as green; if they had access to any other table(s), that access gets revoked"), this appears to be a backend bug.

I start with the following permissions for a particular db/group:

```javascript
{native: "none", schemas: {PUBLIC: {1: "all", 2: "none", 3: "none", 4: "none"}}}
```

In the UI I grant permission to table 3, and hit save. The following is POSTed:

```javascript
{native: "none", schemas: {PUBLIC: {1: "all", 2: "none", 3: "all", 4: "none"}}}
```

And the following is persisted and returned:

```javascript
{native: "none", schemas: {PUBLIC: {1: "none", 2: "none", 3: "all", 4: "none"}}}
```

**Comment by salsakran** [comment]
I think what you're describing sounds correct.

I.e. if I don't have access to the Users table, I should still be able to see the id, but I shouldn't be able to "follow" that link in the query builder. 

I also don't think the front end should be doing permissions checking of this kind.

**Comment by camsaul** [comment]
Yeah I guess that makes sense, I can look at this. 

I think tho if fetching metadata for a related table fails for some reason we should still try to handle that a little more gracefully

**Comment by salsakran** [comment]
breaking that out into a separate issue, and clarifying this one.

**Comment by tlrobinson** [comment]
Yeah, we have a global 403 handler shows the error on any 403 response, which might be a little heavy handed. We started whitelisting certain endpoints (e.x. for dashboard cards) but maybe we should rethink that.

Regardless of that, this is a (minor) permission leak that should be fixed (user normally wouldn't be able to get any info about the target pk)

**Comment by camsaul** [comment]
@tlrobinson what metadata endpoint is this being used here

**Comment by tlrobinson** [comment]
`/api/table/TABLE_ID/query_metadata`, but we should make sure the same problem isn't present on others.

**Comment by camsaul** [comment]
Fixed by #3892
